### PR TITLE
Lock transaction/v1

### PR DESCRIPTION
**Pull request by cryogenian**
[ch13007] [ch13008] 

`Flow` seems to be good candidate for extraction into lib-jdbc. Will extract it after/during postgres udpate.

### CODE REVIEW

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        }
+
+        def append: ConnectionIO[Unit] = {
+          val mbFilter = idColumn traverse_ { col => filterTempIds(col) }
+          mbFilter >>
+          insertInto >>
+          truncate
+        }
+
+        def persist(writeMode: WriteMode): ConnectionIO[Unit] = {
```
Comment (wemrysi) [inline destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: We might consider moving the `WriteMode` argument to the constructor, with that information we could choose to use [temporary tables](https://docs.microsoft.com/en-us/sql/t-sql/statements/create-table-transact-sql?view=sql-server-ver15#temporary-tables) when possible. We'd also be able to fail faster in the `Create` case.

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+          val fragment =
+            fr"DELETE target FROM" ++ tableFragment ++
+            fr" target INNER JOIN" ++ tempFragment ++ fr" temp" ++
+            fr"ON" ++ mkColumn("target") ++ fr0"=" ++ mkColumn("temp")
+
+          fragment.updateWithLogHandler(log).run.void
+        }
+
+        def append: ConnectionIO[Unit] = {
+          val mbFilter = idColumn traverse_ { col => filterTempIds(col) }
```
Comment (wemrysi) [inline destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: I'm not sure this is safe for non-upsert sinks and if it is (i.e. all appends contain new ids), then it shouldn't be necessary.

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        private def rename: ConnectionIO[Unit] =
+          (fr0"EXEC SP_RENAME '" ++
+            tempFragment ++ fr0"', '" ++
+            Fragment.const0(SQLServerHygiene.hygienicIdent(Ident(unsafeName)).unsafeForSqlName) ++ fr0"'")
+            .updateWithLogHandler(log)
+            .run
+            .void
+
+        private def filterTempIds(idColumn: Column[_]): ConnectionIO[Unit] = {
```
Comment (wemrysi) [inline destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala:197]: Might make sense to index the temp table for this join, though we'd want to see some query plans/test it to be sure.

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala (hunk 1 above)
Comment (cryogenian) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: The problem with temp tables is that we _ptobably_ want to resume push even if sql server is restarted which will create different session (if I understand session correctly).

I don't mind moving write mode to constructor, but not sure how would it help. Could you elaborate please?

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala (hunk 2 above)
Comment (cryogenian) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: Right, thank you.

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala (hunk 3 above)
Comment (cryogenian) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala:197]: It wouldn't harm I suppose. Target table has index and temp one should be small.

File: destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala (hunk 1 above)
Comment (wemrysi) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: > we probably want to resume push even if sql server is restarted

Are you referring to a restart during a push or in between pushes? It seems like the only time _during_ would really be significant is during the initial load of an incremental push, but there we'd be performing a replace/create so we wouldn't use a temp table anyhow. For subsequent loads of an incremental, we'd be resuming from previous so is it worth it to try and resume mid push, or just restart it?

As for what I meant about "failing faster", if we know the write mode ahead of time, we could check for the existence of the destination table in the "create" case before consuming anything (whereas now we may consume everything and then fail during `persist` b/c the table already exists).

Comment (cryogenian) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: E.g. during a push connection was reset, we catch the error and retry insert chunk after timeout. Or we don't simply rollback on error during rename but again retry. 

Right, I'll add a check that table exists or not exists. Not sure about this though, as it's not locked and user might delete or create it during loading.

Comment (wemrysi) [reply destination/src/main/scala/quasar/plugin/sqlserver/destination/TempTableFlow.scala]: > during a push connection was reset, we catch the error and retry insert chunk after timeout. Or we don't simply rollback on error during rename but again retry.

Yeah, is that what happens now (we retry)? But you're correct (afaict) that any temp tables would be gone, so we'd have to weigh the benefits.

> it's not locked and user might delete or create it during loading

Yeah, it is mostly to fail-fast in the common case. If it doesn't exist at the start and the user adds a conflicting table before we attempt to create, the create will still fail.

**Review by wemrysi — APPROVED** [review]

**Pull request merged**